firedrake/adjoint/constant.py

```diff
@@ -85,16 +84,14 @@ def _ad_dot(self, other, options=None):
 
     @staticmethod
     def _ad_assign_numpy(dst, src, offset):
-        dst.assign(Constant(numpy.reshape(src[offset:offset + dst.value_size()], dst.ufl_shape)))
-        offset += dst.value_size()
+        l = dst.ufl_element().value_size()
+        dst.assign(numpy.reshape(src[offset:offset + l], dst.ufl_shape))
```

Comment:
This was probably wrong before, but almost certainly this assignment wants to be a in a `with no_annotations` block or however it's spelt.